src/libostree/ostree-repo-pull.c

```diff
@@ -1321,12 +1335,166 @@ commitstate_is_partial (OtPullData   *pull_data,
+}
+#endif  /* OSTREE_ENABLE_EXPERIMENTAL_API */
+
+/* Verify the ref and collection bindings.
+ *
+ * The ref binding is verified only if it exists. But if we have the
+ * collection ID specified in the remote configuration then the ref
+ * binding must exist, otherwise the verification will fail. Parts of
+ * the verification can be skipped by passing NULL to requested_ref or
+ * checksum parameters.
+ *
+ * The collection binding is verified only when we have collection ID
+ * specified in the remote configuration. If it is specified, then the
+ * binding must exist and must be equal to the remote repository
+ * collection ID.
+ */
+static gboolean
+verify_bindings (OtPullData                 *pull_data,
+                 const char                 *checksum,
+                 GVariant                   *commit,
+                 const OstreeCollectionRef  *requested_ref,
+                 GError                    **error)
+{
+  g_autofree char *remote_collection_id = NULL;
+#ifdef OSTREE_ENABLE_EXPERIMENTAL_API
+  remote_collection_id = get_remote_repo_collection_id (pull_data);
+#endif  /* OSTREE_ENABLE_EXPERIMENTAL_API */
+  g_autoptr(GVariant) metadata = g_variant_get_child_value (commit, 0);
+  g_autofree const char **refs = NULL;
+  if (!g_variant_lookup (metadata,
+                         OSTREE_REF_BINDING,
+                         "^a&s",
+                         &refs))
+    {
+      /* Early return here - if the remote collection ID is NULL, then
+       * we certainly will verify the collection binding in the
+       * commit.
+       */
+      if (remote_collection_id == NULL)
+        return TRUE;
+
+      return glnx_throw (error,
+                         "expected commit metadata to have ref "
+                         "binding information, found none");
+    }
+
+  if (requested_ref != NULL)
+    {
+      if (!g_strv_contains ((const char *const *) refs, requested_ref->ref_name))
+        {
+          g_autofree char *refs_str = NULL;
+
+          if (refs != NULL && (*refs) != NULL)
+            {
+              g_autoptr(GString) refs_dump = g_string_new (NULL);
+
+              for (const char **iter = refs; *iter != NULL; ++iter)
+                {
+                  const char *ref = *iter;
+
+                  g_string_append_printf (refs_dump, "‘%s’, ", ref);
+                }
+              g_assert_cmpint (refs_dump->len, >=, 2);
+              gsize stray_comma = refs_dump->len - 2;
+              refs_str = g_string_free (g_steal_pointer (&refs_dump), FALSE);
+              refs_str[stray_comma] = '\0';
+            }
+          else
+            {
+              refs_str = g_strdup ("no refs");
+            }
+
+          return glnx_throw (error, "commit has no requested ref ‘%s’ "
+                             "in ref binding metadata (%s)",
+                             requested_ref->ref_name, refs_str);
+        }
+
+      if (checksum != NULL)
```

Comment:
That makes sense to me. I guess the refs could get swapped if `opt_collection_refs_set` were used incorrectly, so this check is relevant. I’m not sure I can see how it could affect the results of a pull which uses the ref → checksum mapping from the `summary` file, since the refs can’t get swapped that way. I’ll add a comment to the code.